### ISSUE TITLE
Rule Docs do not provide guidance for no-mutating-props, when mutating a property of a prop

### ISSUE THREAD

**Issue by douglasg14b** [vuejs/eslint-plugin-vue#1772]
This rule will flag when mutating a value on a prop, however, the docs don't show what the expected solution is. Take the following example:

```vue
<template>
  <div>
    <input v-model="person.age">
  </div>
</template>
<script>
  export default {
    props: {
      person: {
        type: Object,
        required: true
      }
    },
  }
</script>
```

This results in an error:

> Unexpected mutation of "person" prop.

However, the rule docs only provide examples for mutating the prop directly, not for mutating a value of an object.

* https://eslint.vuejs.org/rules/no-mutating-props.html
* https://vuejs.org/v2/style-guide/index.html#Implicit-parent-child-communication-use-with-caution

The closest the Vue docs have is: ` @input="$emit('input', $event.target.value)"`, however, if `$event` is just the new value there is no `target` nor a `target.value`. **What, *exact*, event is expected to be emitted here if done manually?**

**Comment by douglasg14b** [comment]
Thanks for the reply!

<hr>


> Thank you for posting this issue. but, I don't know what you are looking for.

The example above is shown as an error, the lint rule docs provide no guidance on how that error is expected to be resolved. I am asking how it is expected to be resolved. This is a linting plugin, which is enforcing opinions/standards, what opinion/standard is it enforcing, what is an example that solves for that opinion/standard?

> If you find the document confusing, feel free to open a pull request to improve the document.

I am asking because I do not know, there is nothing for me to add to the document. The purpose of this issue is to solicit such information from the authors.


> I don't know what event you need to emit or not. Because it depends on your component and the version of Vue you use.

I have provided an example component in the OP, I did however fail to provide a Vue version. This is Vue 2. What is expected to be emitted is an event, as per the current recommendations for this rule and in the vue docs (Which are linked above). However, those recommendations don't provide guidance for this example.

**Comment by ota-meshi** [comment]
> I have provided an example component in the OP, I did however fail to provide a Vue version. This is Vue 2.

I don't think you need to take `person` as a props because your component only uses the `age` of the `person` props.

```vue
<template>
  <div>
    <input :value="value" @input="changeAge">
  </div>
</template>
<script>
  export default {
    props: {
      value: {
        type: string
      }
    },
    methods: {
        changeAge (e) {
            this.$emit('input',  e.target.value)
        }
    }
  }
</script>
```

Parent component that use it can use `v-model`.

```vue
<YourComponent v-model="person.age" />
```

If the first component is simplified and you want to keep the `person` props, you can do the following:

```vue
<template>
  <div>
    <input :value="person.age" @input="changeAge">
  </div>
</template>
<script>
  export default {
    props: {
      person: {
        type: Object,
        required: true
      }
    },
    methods: {
        changeAge (e) {
            this.$emit('update:person', { ...this.person, age: e.target.value})
        }
    }
  }
</script>
```

The parent component is:

```vue
<YourComponent v-bind:person.sync="person" />
```

If you want to use v-model:

```vue
<template>
  <div>
    <input :value="person.age" @input="changeAge">
  </div>
</template>
<script>
  export default {
    model: {
      prop: 'person'
    },
    props: {
      person: {
        type: Object,
        required: true
      }
    },
    methods: {
        changeAge (e) {
            this.$emit('update:person', { ...this.person, age: e.target.value})
        }
    }
  }
</script>
```

The parent component is:

```vue
<YourComponent v-model="person" />
```

https://vuejs.org/v2/guide/components-props.html
https://vuejs.org/v2/guide/components-custom-events.html
